taiga_halo2/src/circuit/schnorr.rs

```diff
@@ -0,0 +1,398 @@
+        }
+    }
+
+    #[allow(non_snake_case)]
+    fn synthesize(
+        &self,
+        config: Self::Config,
+        mut layouter: impl Layouter<pallas::Base>,
+    ) -> Result<(), plonk::Error> {
+        SinsemillaChip::<
+                NoteCommitmentHashDomain,
+                NoteCommitmentDomain,
+                NoteCommitmentFixedBases,
+            >::load(config.sinsemilla_config, &mut layouter)?;
+        // We implement the verification algorithm first
+        // and assume that the signature is given
+        // Construct an ECC chip
+        let ecc_chip = EccChip::construct(config.ecc_config);
+        let zero_cell = assign_free_advice(
+            layouter.namespace(|| "zero"),
+            config.advices[0],
+            Value::known(Fp::zero()),
+        )?;
+        // TODO: Message length (256bits) is bigger than the size of Fp (255bits)
+        // Obtain message: m
+        let m_cell = assign_free_instance(
+            layouter.namespace(|| "message instance"),
+            config.primary,
+            0,
+            config.advices[0],
+        )
+        .unwrap();
+        // Obtain the signature: (r,s)
+        let r_cell = {
+            let (r, _, _) = self.r.jacobian_coordinates();
+            assign_free_advice(
+                layouter.namespace(|| "r"),
+                config.advices[0],
+                Value::known(r),
+            )?
+        };
+        let s_scalar = ScalarFixed::new(
+            ecc_chip.clone(),
+            layouter.namespace(|| "s"),
+            Value::known(self.s),
+        )?;
+        // Obtain public key : P
+        let p_cell = {
+            let (p, _, _) = self.pk.jacobian_coordinates();
+            assign_free_advice(
+                layouter.namespace(|| "p"),
+                config.advices[0],
+                Value::known(p),
+            )?
+        };
+
+        // Verify: s*G = R + Hash(r||P||m)*P
+        // s*G
+        let generator = FixedPoint::from_inner(ecc_chip.clone(), NoteCommitmentFixedBasesFull);
+        let (sG, _) = generator.mul(layouter.namespace(|| "s_scalar * generator"), &s_scalar)?;
+
+        // Hash(r||P||m)
+        let h_scalar = {
+            let poseidon_chip = PoseidonChip::construct(config.poseidon_config);
+            let poseidon_message = [r_cell, p_cell, m_cell, zero_cell];
+            let poseidon_hasher =
+                PoseidonHash::<_, _, poseidon::P128Pow5T3, ConstantLength<4>, 3, 2>::init(
+                    poseidon_chip,
+                    layouter.namespace(|| "Poseidon init"),
+                )?;
+            let h = poseidon_hasher.hash(
+                layouter.namespace(|| "Poseidon_hash(r, P, m)"),
+                poseidon_message,
+            )?;
+
+            ScalarVar::from_base(
+                ecc_chip.clone(),
+                layouter.namespace(|| "ScalarVar from_base"),
+                &h,
+            )?
+        };
+
+        let R = NonIdentityPoint::new(
```

Comment:
The `R` should be the same as the above `r`. And if we assign(NonIdentityPoint::new) it twice, there might be some problems.